chia/daemon/server.py

```diff
@@ -561,6 +566,58 @@ async def get_routes(self, websocket: WebSocketResponse, request: Dict[str, Any]
         response: Dict[str, Any] = {"success": True, "routes": routes}
         return response
 
+    async def get_wallet_addresses(self, websocket: WebSocketResponse, request: Dict[str, Any]) -> Dict[str, Any]:
+        all_keys = Keychain().get_keys(include_secrets=True)
+        fingerprints = request.get("fingerprints", None)
+        index = request.get("index", 0)
+        count = request.get("count", 1)
+        non_observer_derivation = request.get("non_observer_derivation", False)
+
+        # if fingerprints is None, we want all keys, otherwise we want the keys that match the fingerprints
+        if fingerprints is None:
+            keys = all_keys
+        else:
+            keys_by_fingerprint = {key.fingerprint: key for key in all_keys}
+            keys = []
+            missing_fingerprints = set()
+            for fingerprint in fingerprints:
+                if fingerprint not in keys_by_fingerprint:
+                    missing_fingerprints.add(fingerprint)
+                else:
+                    keys.append(keys_by_fingerprint[fingerprint])
+
+            if len(keys) != len(fingerprints):
+                return {"success": False, "error": f"key(s) not found for fingerprint(s) {missing_fingerprints}"}
+
+        selected = self.net_config["selected_network"]
+        prefix = self.net_config["network_overrides"]["config"][selected]["address_prefix"]
+
+        wallet_addresses_by_fingerprint = {}
+        for key in keys:
+            address_entries = []
+
+            # we require access to the private key to generate wallet addresses
```

Comment:
this isnt true for unhardened tho is it?

Reply:
Well yes and no... If given the wallet's root public key, you're right, we should be able to derive the unhardened addresses from that wallet root public key. But we don't store public keys today, and the `master_sk_to_wallet_sk_unhardened()` function currently assumes that the master private key is provided, so we'd have to make more sweeping changes to support observer mode.